db/versions/0091-migration.sql

```diff
@@ -0,0 +1,166 @@
+  --   WHERE queue_name NOT IN ('claim-queue', 'deadline-queue', 'resolved-queue')
+  --   AND task_queue_id IS NULL
+  --   LIMIT 1
+  -- ) THEN
+  --   RAISE EXCEPTION 'Not possible to migrate, some records are still in old format';
+  -- END IF;
+
+  -- migration of data
+  -- prevent reads and writes from table, get exclusive access
+  LOCK TABLE azure_queue_messages;
+
+  -- Task Deadlines
+  -- purpose: know when particular task expires, so dependencies/task groups can be resolved
+  CREATE TABLE queue_task_deadlines (
+    task_id text not null,
+    task_group_id text not null,
+    scheduler_id text not null,
+    created timestamptz not null,
+    deadline timestamptz not null,
+    visible timestamptz not null,
+    pop_receipt uuid null
+  );
+
+  -- migrate data to deadline queue
+  INSERT INTO queue_task_deadlines
+    (task_group_id, task_id, scheduler_id, created, deadline, visible, pop_receipt)
+  SELECT
+    convert_from(decode(message_text, 'base64'), 'utf-8')::jsonb->>'taskGroupId',
+    convert_from(decode(message_text, 'base64'), 'utf-8')::jsonb->>'taskId',
+    convert_from(decode(message_text, 'base64'), 'utf-8')::jsonb->>'schedulerId',
+    inserted,
+    CAST(convert_from(decode(message_text, 'base64'), 'utf-8')::jsonb->>'deadline' AS timestamp with time zone),
+    visible, -- when this message becomes visible, it means it hits deadline
+    pop_receipt
+  FROM azure_queue_messages
+  WHERE queue_name = 'deadline-queue'
+  AND expires > now();
+
+  CREATE INDEX queue_task_deadline_idx ON queue_task_deadlines (task_id);
+  CREATE INDEX queue_task_deadline_vis_idx ON queue_task_deadlines (visible);
+
+  -- Resolved Tasks
+  -- queries: what tasks are not scheduled for given task_group_id/scheduler_id
+  -- purpose: not much since this will have short-lived data
+  CREATE TABLE queue_resolved_tasks (
+    task_group_id text not null,
+    task_id text not null,
+    scheduler_id text not null,
+    resolution text not null,
+    resolved timestamptz not null,
+    visible timestamptz not null,
+    pop_receipt uuid null
+  );
+
+  -- migrate data to resolved queue
+  INSERT INTO
+    queue_resolved_tasks (task_id, task_group_id, scheduler_id, resolution, resolved, visible, pop_receipt)
+  SELECT
+    convert_from(decode(message_text, 'base64'), 'utf-8')::jsonb->>'taskId',
+    convert_from(decode(message_text, 'base64'), 'utf-8')::jsonb->>'taskGroupId',
+    convert_from(decode(message_text, 'base64'), 'utf-8')::jsonb->>'schedulerId',
+    convert_from(decode(message_text, 'base64'), 'utf-8')::jsonb->>'resolution',
+    inserted,
+    visible,
+    pop_receipt
+  FROM azure_queue_messages
+  WHERE queue_name = 'resolved-queue'
+  AND expires > now();
+
+  CREATE INDEX queue_resolved_task_idx ON queue_resolved_tasks (task_id);
+  CREATE INDEX queue_resolved_task_vis_idx ON queue_resolved_tasks (visible);
+
+  -- Claimed Tasks
+  -- queries: what tasks are running and waiting to be reclaimed or resolved
```

Comment:
This is good, and I believe visibility here is based on the reclaim time?

Reply:
Yes, this is the same value, until the records is fetched for processing. 
We could probably still use `taken_until` to act as `visible` column, and update it, to temporarily hide it during processing